apps/src/studio/studio.js

```diff
@@ -4753,71 +4754,78 @@ Studio.allGoalsVisited = function() {
-  }
+    if (valueName == 'currentPointsAtOrAbove' && Studio.playerScore < value) {
+      return false;
+    }
 
-  if (required.setBotMap && !tracked.hasSetMap) {
-    return { exists: true, achieved: false, message: studioMsg.failedHasSetMap() };
-  }
+    if (valueName == 'currentPointsBelow' && Studio.playerScore >= value) {
+      return false;
+    }
 
-  if (required.touchAllItems && Studio.items.length > 0) {
-    return { exists: true, achieved: false, message: studioMsg.failedTouchAllItems() };
-  }
+    if (valueName == 'allGoalsVisited' && Studio.allGoalsVisited() != value) {
+      return false;
+    }
 
-  if (required.scoreMinimum && Studio.playerScore < required.scoreMinimum) {
-    return { exists: true, achieved: false, message: studioMsg.failedScoreMinimum() };
-  }
+    if (valueName == 'setMap' && tracked.hasSetMap != value) {
+      return false;
+    }
 
-  if (required.addItem && !tracked.hasAddedItem) {
-    return { exists: true, achieved: false, message: studioMsg.failedAddItem() };
+    if (valueName == 'setBotSpeed' && tracked.hasSetBotSpeed != value) {
+      return false;
+    }
   }
 
-  if (required.removedItemCount && tracked.removedItemCount < required.removedItemCount) {
-    return { exists: true, achieved: false, message: studioMsg.failedRemovedItemCount() };
-  }
+  return true;
+};
 
-  if (required.winGame && ! tracked.hasWonGame) {
-    return { exists: true, achieved: false, message: studioMsg.failedHasWonGame() };
+/**
+ * A level can provide zero or more progress conditions which are special cases
+ * that we test to see if the level has succeeded or failed.  This function
+ * evaluates the state of these criteria.  It returns false if none of the
+ * criteria affects progress, otherwise an object that contains information
+ * about the specific succeeding or failing criteria.
+ *
+ * @param {Array} conditions. 
+ * @returns {Object|false} outcome
+ * @returns {boolean} outcome.success Whether level has succeeded.
+ * @returns {message} outcome.message Optional result message.
+ */
+Studio.checkProgressConditions = function() {
+  if (!level.progressConditions) {
+    return;
   }
 
-  if (required.lostGame && ! tracked.hasLostGame) {
-    return { exists: true, achieved: false, message: studioMsg.failedHasLostGame() };
-  }
+  for (var i = 0; i < level.progressConditions.length; i++) {
+    var condition = level.progressConditions[i];
 
-  if (required.setActivity &&
-      !(tracked.setActivityRecord && 
-        tracked.setActivityRecord[required.setActivity.itemType] &&
-        tracked.setActivityRecord[required.setActivity.itemType][required.setActivity.activityType])) {
-    return { exists: true, achieved: false, message: studioMsg.failedSetActivity() };
+    if (Studio.conditionSatisfied(condition.required)) {
+      return condition.result;
+    }
   }
 
-  return { exists: true, achieved: true, message: null };
+  return false;
```

Comment:
You have inconsistent return types in this method.  The documented result type is `{false|{success: boolean, message: message}}` but you also return `undefined` in your early-out case.  I'd like to see the early-out and fall-through case return the same thing, and I think `undefined` or `null` makes sense, in the "no outcome yet" sense.

Might be good to 
```javascript
/**
 * @typedef {Object} StudioOutcome
 * @property {boolean} success
 * @property {string} message
 */
```